File: app/src/main/java/com/marknkamau/justjava/utils/RealmUtils.java
package com.marknkamau.justjava.utils;

import com.marknkamau.justjava.models.CartItem;

import io.realm.Realm;
import io.realm.RealmResults;
import timber.log.Timber;

public class RealmUtils {
    private Realm realm;
    private RealmResults<CartItem> allCartItems;

    public RealmUtils() {
        realm = Realm.getDefaultInstance();
    }

    public RealmResults<CartItem> getAllCartItems() {
        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                allCartItems = realm.where(CartItem.class).findAll();
            }
        });
        return allCartItems;
    }

    public void saveNewItem(final CartItem cartItem) {
        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                Number current = realm.where(CartItem.class).max("itemID");
                int nextID;
                if (current == null) {
                    nextID = 1;
                } else {
                    nextID = current.intValue() + 1;
                }
                cartItem.setItemID(nextID);

                realm.copyToRealm(cartItem);
            }
        });
    }

    public void deleteAllItems() {
        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                realm.deleteAll();
            }
        });
    }

    public int getTotalCost() {
        return realm.where(CartItem.class).sum("itemPrice").intValue();
    }

    public void deleteSingleItem(final CartItem item) {
        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                item.deleteFromRealm();
            }
        });
    }

    public void saveEdit(final CartItem item) {
        Timber.i("Changing...");
        realm.executeTransaction(new Realm.Transaction() {
            @Override
            public void execute(Realm realm) {
                realm.copyToRealmOrUpdate(item);
            }
        });
    }

}

File: app/src/main/java/com/marknkamau/justjava/utils/RealmUtils.kt
package com.marknkamau.justjava.utils

import com.marknkamau.justjava.models.CartItem

import io.realm.Realm
import io.realm.RealmResults
import timber.log.Timber

class RealmUtils {
    private val realm: Realm = Realm.getDefaultInstance()
    private lateinit var allCartItems: RealmResults<CartItem>

    fun getAllCartItems(): RealmResults<CartItem> {
        realm.executeTransaction { realm -> allCartItems = realm.where(CartItem::class.java).findAll() }
        return allCartItems
    }

    fun saveNewItem(cartItem: CartItem) {
        realm.executeTransaction { realm ->
            val current = realm.where(CartItem::class.java).max("itemID")
            val nextID: Int
            if (current == null) {
                nextID = 1
            } else {
                nextID = current.toInt() + 1
            }
            cartItem.itemID = nextID

            realm.copyToRealm(cartItem)
        }
    }

    fun deleteAllItems() {
        realm.executeTransaction { realm -> realm.deleteAll() }
    }

    val totalPrice: Int
        get() = realm.where(CartItem::class.java).sum("itemPrice").toInt()

    fun deleteSingleItem(item: CartItem) {
        realm.executeTransaction { item.deleteFromRealm() }
    }

    fun saveEdit(item: CartItem) {
        realm.executeTransaction { realm -> realm.copyToRealmOrUpdate(item) }
    }

}
